File: m4d/ClientApp/src/models/DanceRatingDelta.ts
export class DanceRatingDelta {
  public static fromString(value: string): DanceRatingDelta {
    if (!value) {
      throw new Error("value must not be falsey");
    }

    const groups = /([A-Z0-9]+)([+-]\d+)/g.exec(value);

    if (groups?.length !== 3) {
      throw new Error(`value "${value}" must conform to {dancid}(+|-){count}`);
    }

    return new DanceRatingDelta(groups[1], Number.parseInt(groups[2], 10));
  }

  constructor(
    readonly danceId: string,
    readonly delta: number,
  ) {}

  public toString(): string {
    return `${this.danceId}${this.delta < 0 ? "-" : "+"}${Math.abs(this.delta)}`;
  }
}

export enum VoteDirection {
  Up,
  Down,
}

export class DanceRatingVote {
  constructor(
    readonly danceId: string,
    readonly direction: VoteDirection,
  ) {}
}

File: m4dModels/DanceRatingDelta.cs
using System;

namespace m4dModels
{
    public class DanceRatingDelta
    {
        public DanceRatingDelta()
        {
        }

        public DanceRatingDelta(string id, int delta)
        {
            DanceId = id;
            Delta = delta;
        }

        public DanceRatingDelta(string value)
        {
            ArgumentNullException.ThrowIfNull(value);

            var parts = value.Split('+', '-');

            var sign = value.Contains('-') ? -1 : 1;
            var offset = 1;

            DanceId = parts[0];
            if (parts.Length > 1)
            {
                _ = int.TryParse(parts[1], out offset);
            }

            Delta = sign * offset;
        }

        public override string ToString()
        {
            return $"{DanceId}{(Delta < 0 ? "-" : "+")}{Math.Abs(Delta)}";
        }

        public string DanceId { get; set; }
        public int Delta { get; set; }
    }
}
